### ISSUE TITLE
BeEquivalentTo fails with default interface methods (6.12.1)

### ISSUE THREAD

**Issue by dusk0r** [fluentassertions/fluentassertions#2759]
### Description

Comparing the same Interface with BeEquivalentTo() fails if the interface contains default interface methods. 

### Reproduction Steps

```csharp
    public void BeEquivalentTo_Default_Interface_Method()
    {
        List<ITest> lista = [new Test { Name = "Test" }];
        List<ITest> listb = [new Test { Name = "Test" }];

        lista.Should().BeEquivalentTo(listb); // Fails with: Expectation has property lista[0].NameLength that the other object does not have.
    }
    
    public class Test : ITest
    {
        public string Name { get; set; }
    }

    public interface ITest
    {
        public string Name { get; }
        public int NameLength => Name.Length;
    }
```

### Expected behavior

Doesn't thorw an exception.

### Actual behavior

Fails with: Expectation has property lista[0].NameLength that the other object does not have.

### Regression?

This worked in 6.12.0 and before but fails in 6.12.1.

### Configuration

Fluent Assertion 6.12.1
.net 8.0.401

### Are you willing to help with a pull-request?

No

**Comment by ITaluone** [comment]
I don't think this is the exception message (`Expectation has property lista[0].NameLength that the other object does not have.`), because this works only with v7 (meaning the PR isn't even merged yet)..

**Comment by ITaluone** [comment]
Sounds like #2754

**Comment by omarjuul** [comment]
I also ran into this issue when updating to v6.12.1 today. I just finished creating what I thought was a minimal repro when I saw this issue with a smaller repro.
I don't see the correlation with #2754 as that issue seems to be about priority, whereas this issue is about being able to find the property at all.

**Comment by dennisdoomen** [comment]
I'm struggling to fix this problem and wondering whether we should at all. 

The problem is that during the nested equivalency check, FA doesn't actually now the compile-time type of the `Test` instance and only sees the `Test` class. Default interface properties are only visible through the interface and not the type that implements the interface. This kind of worked because the 6.12 implementation was a bit naive. Fixing this will be hard, since we changed the implementation considerably to fix #2146. 

The only way to fix this is to enhance the mechanism that scans for the members of a type, to also include interfaces they implement. But that will break some of the more complicated scenarios around explicitly implemented properties.

**Comment by MartinDembergerR9** [comment]
Sorry, but I don't understand your argumentation.
Isn't the case here that the members of the elements of lista are determined by the compile time type and the members of listb by the runtime type?

**Comment by dennisdoomen** [comment]
> Isn't the case here that the members of the elements of lista are determined by the compile time type and the members of listb by the runtime type?

Exactly my point. And since `Test` doesn't have `NameLength` but `ITest` does, and we don't know that the items of `listb` are of type `ITest`, we don't "see" `NameLenght`. I quickly tried an implementation to include all members of all interfaces that `Test` implements, but that broke other tests.

**Comment by dennisdoomen** [comment]
Resolved in 6.12.2
